File: src/omnisus_db/sources/datasus_ftp/dbc.py
"""DATASUS DBC: a DBF header, a CRC32, then a PKWare DCL imploded body.

The decoder below is an altered port of Mark Adler's blast.c (zlib
contrib/blast, version 1.3). Altered: it reads an in-memory buffer, appends to a
growing output instead of a 4 KiB window, and raises InvalidDbcError instead of
returning codes. ``native/omnisus-db-dbf/src/dbc.rs`` carries the same port.
Original notice, from blast.h:

    Copyright (C) 2003, 2012, 2013 Mark Adler

    This software is provided 'as-is', without any express or implied
    warranty.  In no event will the author be held liable for any damages
    arising from the use of this software.

    Permission is granted to anyone to use this software for any purpose,
    including commercial applications, and to alter it and redistribute it
    freely, subject to the following restrictions:

    1. The origin of this software must not be misrepresented; you must not
       claim that you wrote the original software. If you use this software
       in a product, an acknowledgment in the product documentation would be
       appreciated but is not required.
    2. Altered source versions must be plainly marked as such, and must not be
       misrepresented as being the original software.
    3. This notice may not be removed or altered from any source distribution.
"""

from __future__ import annotations

import structlog

from omnisus_db.sources.datasus_ftp.native import Backend, load_native, requested_backend

logger = structlog.get_logger(__name__)

HINT_BYTES = 16 * 1024**2
"""Compressed size above which pure Python (about 10 MB of output per second) earns a hint."""

_MAXBITS = 13


class InvalidDbcError(ValueError):
    """The DBC payload is malformed or truncated."""


def _huffman(compact: bytes) -> tuple[list[int], list[int]]:
    """Canonical decoding tables (codes per length, symbols by length) from compact lengths.

    Each compact byte is a code length (low four bits) repeated (high four bits + 1) times.
    """
    lengths: list[int] = []
    for byte in compact:
        lengths += [byte & 15] * ((byte >> 4) + 1)
    count = [0] * (_MAXBITS + 1)
    for length in lengths:
        count[length] += 1
    offsets = [0] * (_MAXBITS + 1)
    for length in range(1, _MAXBITS):
        offsets[length + 1] = offsets[length] + count[length]
    symbols = [0] * len(lengths)
    for symbol, length in enumerate(lengths):
        if length:
            symbols[offsets[length]] = symbol
            offsets[length] += 1
    return count, symbols


_LITERAL = _huffman(
    bytes(
        [
            11, 124, 8, 7, 28, 7, 188, 13, 76, 4, 10, 8, 12, 10, 12, 10, 8, 23, 8,
            9, 7, 6, 7, 8, 7, 6, 55, 8, 23, 24, 12, 11, 7, 9, 11, 12, 6, 7, 22, 5,
            7, 24, 6, 11, 9, 6, 7, 22, 7, 11, 38, 7, 9, 8, 25, 11, 8, 11, 9, 12,
            8, 12, 5, 38, 5, 38, 5, 11, 7, 5, 6, 21, 6, 10, 53, 8, 7, 24, 10, 27,
            44, 253, 253, 253, 252, 252, 252, 13, 12, 45, 12, 45, 12, 61, 12, 45,
            44, 173,
        ]
    )
)  # fmt: skip
_LENGTH = _huffman(bytes([2, 35, 36, 53, 38, 23]))
_DISTANCE = _huffman(bytes([2, 20, 53, 230, 247, 151, 248]))
_LENGTH_BASE = (3, 2, 4, 5, 6, 7, 8, 9, 10, 12, 16, 24, 40, 72, 136, 264)
_LENGTH_EXTRA = (0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8)
_END = 519


def _explode(data: bytes, start: int) -> bytes:
    """Decode the imploded stream at ``data[start:]``; errors cite absolute offsets."""
    position = start
    buffer = 0
    available = 0

    def bits(need: int) -> int:
        # Bits are packed least significant first.
        nonlocal position, buffer, available
        while available < need:
            if position == len(data):
                raise InvalidDbcError(f"truncated DBC stream at byte {position}")
            buffer |= data[position] << available
            position += 1
            available += 8
        value = buffer & ((1 << need) - 1)
        buffer >>= need
        available -= need
        return value

    def decode(table: tuple[list[int], list[int]]) -> int:
        # Codes are stored bit-reversed and inverted relative to canonical order.
        count, symbols = table
        code = first = index = 0
        for length in range(1, _MAXBITS + 1):
            code |= bits(1) ^ 1
            if code < first + count[length]:
                return symbols[index + code - first]
            index += count[length]
            first = (first + count[length]) << 1
            code <<= 1
        raise InvalidDbcError(f"corrupt DBC stream at byte {position}")

    literals_coded = bits(8)
    if literals_coded > 1:
        raise InvalidDbcError(f"corrupt DBC stream at byte {position}")
    dictionary_bits = bits(8)
    if dictionary_bits not in (4, 5, 6):
        raise InvalidDbcError(f"corrupt DBC stream at byte {position}")
    out = bytearray()
    while True:
        if not bits(1):
            out.append(decode(_LITERAL) if literals_coded else bits(8))
            continue
        symbol = decode(_LENGTH)
        length = _LENGTH_BASE[symbol] + bits(_LENGTH_EXTRA[symbol])
        if length == _END:
            return bytes(out)
        shift = 2 if length == 2 else dictionary_bits
        distance = (decode(_DISTANCE) << shift) + bits(shift) + 1
        if distance > len(out):
            raise InvalidDbcError(f"corrupt DBC stream at byte {position}")
        source = len(out) - distance
        if distance >= length:
            out += out[source : source + length]
        else:
            # Overlapping copy: each new byte may be one this copy just wrote.
            for index in range(source, source + length):
                out.append(out[index])


def _python_decompress(raw: bytes) -> bytes:
    if len(raw) < 10:
        raise InvalidDbcError("missing DBC header")
    header_size = int.from_bytes(raw[8:10], "little")
    if header_size + 4 > len(raw):
        raise InvalidDbcError("DBC header size exceeds file size")
    # The 4 bytes after the header are a CRC32 that DATASUS readers ignore.
    return raw[:header_size] + _explode(raw, header_size + 4)


def decompress_bytes(raw: bytes, backend: Backend | None = None) -> bytes:
    """DBF bytes from a DBC payload.

    ``backend``, else ``OMNISUS_DBC_BACKEND``, is ``python``, ``rust`` or ``auto``
    (the native package when installed). Both produce identical bytes.

    Raises:
        InvalidDbcError: the payload is malformed or truncated.
    """
    requested = requested_backend(backend, variable="OMNISUS_DBC_BACKEND", label="DBC")
    native = load_native(requested)
    logger.debug(
        "datasus_ftp.dbc_backend",
        backend="python" if native is None else "rust",
        requested=requested,
        version=None if native is None else native.__version__,
        fallback="extension_not_installed" if native is None and requested == "auto" else None,
    )
    if native is not None:
        try:
            return native.decompress_dbc(raw)
        except native.InvalidDbcError as exc:
            raise InvalidDbcError(str(exc)) from exc
    if requested == "auto" and len(raw) > HINT_BYTES:
        logger.info("datasus_ftp.dbc_python_backend_slow", compressed_bytes=len(raw))
    return _python_decompress(raw)

File: native/omnisus-db-dbf/src/dbc.rs
//! DATASUS DBC: a DBF header, a CRC32, then a PKWare DCL imploded body.
//!
//! The decoder is an altered port of Mark Adler's blast.c (zlib contrib/blast,
//! version 1.3). Altered: it reads a slice, appends to a growing `Vec` instead
//! of a 4 KiB window, and returns `DbcError` instead of integer codes.
//! `src/omnisus_db/sources/datasus_ftp/dbc.py` carries the same port; both must
//! produce identical bytes and messages. Original notice, from blast.h:
//!
//!   Copyright (C) 2003, 2012, 2013 Mark Adler
//!
//!   This software is provided 'as-is', without any express or implied
//!   warranty.  In no event will the author be held liable for any damages
//!   arising from the use of this software.
//!
//!   Permission is granted to anyone to use this software for any purpose,
//!   including commercial applications, and to alter it and redistribute it
//!   freely, subject to the following restrictions:
//!
//!   1. The origin of this software must not be misrepresented; you must not
//!      claim that you wrote the original software. If you use this software
//!      in a product, an acknowledgment in the product documentation would be
//!      appreciated but is not required.
//!   2. Altered source versions must be plainly marked as such, and must not be
//!      misrepresented as being the original software.
//!   3. This notice may not be removed or altered from any source distribution.

use std::{fmt, sync::LazyLock};

#[derive(Debug, PartialEq, Eq)]
pub enum DbcError {
    MissingHeader,
    HeaderSize,
    Corrupt(usize),
    Truncated(usize),
}

impl fmt::Display for DbcError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::MissingHeader => f.write_str("missing DBC header"),
            Self::HeaderSize => f.write_str("DBC header size exceeds file size"),
            Self::Corrupt(at) => write!(f, "corrupt DBC stream at byte {at}"),
            Self::Truncated(at) => write!(f, "truncated DBC stream at byte {at}"),
        }
    }
}

const MAXBITS: usize = 13;
const END: usize = 519;
const LENGTH_BASE: [usize; 16] = [3, 2, 4, 5, 6, 7, 8, 9, 10, 12, 16, 24, 40, 72, 136, 264];
const LENGTH_EXTRA: [u32; 16] = [0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8];

/// Canonical decoding tables: codes per length, and symbols ordered by length.
struct Huffman {
    count: [usize; MAXBITS + 1],
    symbols: Vec<usize>,
}

impl Huffman {
    /// Each compact byte is a code length (low four bits) repeated (high four bits + 1) times.
    fn new(compact: &[u8]) -> Self {
        let lengths: Vec<usize> = compact
            .iter()
            .flat_map(|&byte| {
                std::iter::repeat_n(usize::from(byte & 15), usize::from(byte >> 4) + 1)
            })
            .collect();
        let mut count = [0; MAXBITS + 1];
        for &length in &lengths {
            count[length] += 1;
        }
        let mut offsets = [0; MAXBITS + 1];
        for length in 1..MAXBITS {
            offsets[length + 1] = offsets[length] + count[length];
        }
        let mut symbols = vec![0; lengths.len()];
        for (symbol, &length) in lengths.iter().enumerate() {
            if length != 0 {
                symbols[offsets[length]] = symbol;
                offsets[length] += 1;
            }
        }
        Self { count, symbols }
    }
}

static LITERAL: LazyLock<Huffman> = LazyLock::new(|| {
    Huffman::new(&[
        11, 124, 8, 7, 28, 7, 188, 13, 76, 4, 10, 8, 12, 10, 12, 10, 8, 23, 8, 9, 7, 6, 7, 8, 7, 6,
        55, 8, 23, 24, 12, 11, 7, 9, 11, 12, 6, 7, 22, 5, 7, 24, 6, 11, 9, 6, 7, 22, 7, 11, 38, 7,
        9, 8, 25, 11, 8, 11, 9, 12, 8, 12, 5, 38, 5, 38, 5, 11, 7, 5, 6, 21, 6, 10, 53, 8, 7, 24,
        10, 27, 44, 253, 253, 253, 252, 252, 252, 13, 12, 45, 12, 45, 12, 61, 12, 45, 44, 173,
    ])
});
static LENGTH: LazyLock<Huffman> = LazyLock::new(|| Huffman::new(&[2, 35, 36, 53, 38, 23]));
static DISTANCE: LazyLock<Huffman> =
    LazyLock::new(|| Huffman::new(&[2, 20, 53, 230, 247, 151, 248]));

/// Least-significant-bit-first reader over a slice; offsets are absolute.
struct Bits<'a> {
    data: &'a [u8],
    position: usize,
    buffer: u32,
    available: u32,
}

impl Bits<'_> {
    fn take(&mut self, need: u32) -> Result<usize, DbcError> {
        while self.available < need {
            let byte = *self
                .data
                .get(self.position)
                .ok_or(DbcError::Truncated(self.position))?;
            self.buffer |= u32::from(byte) << self.available;
            self.position += 1;
            self.available += 8;
        }
        let value = self.buffer & ((1 << need) - 1);
        self.buffer >>= need;
        self.available -= need;
        Ok(value as usize)
    }

    /// Codes are stored bit-reversed and inverted relative to canonical order.
    fn decode(&mut self, table: &Huffman) -> Result<usize, DbcError> {
        let (mut code, mut first, mut index) = (0, 0, 0);
        for length in 1..=MAXBITS {
            code |= self.take(1)? ^ 1;
            let count = table.count[length];
            if code < first + count {
                return Ok(table.symbols[index + code - first]);
            }
            index += count;
            first = (first + count) << 1;
            code <<= 1;
        }
        Err(DbcError::Corrupt(self.position))
    }
}

/// Decode the imploded stream at `data[start..]`.
pub fn explode(data: &[u8], start: usize) -> Result<Vec<u8>, DbcError> {
    let mut bits = Bits {
        data,
        position: start,
        buffer: 0,
        available: 0,
    };
    let literals_coded = bits.take(8)?;
    if literals_coded > 1 {
        return Err(DbcError::Corrupt(bits.position));
    }
    let dictionary_bits = bits.take(8)?;
    if !(4..=6).contains(&dictionary_bits) {
        return Err(DbcError::Corrupt(bits.position));
    }
    let mut out = Vec::new();
    loop {
        if bits.take(1)? == 0 {
            let literal = if literals_coded == 1 {
                bits.decode(&LITERAL)?
            } else {
                bits.take(8)?
            };
            out.push(literal as u8);
            continue;
        }
        let symbol = bits.decode(&LENGTH)?;
        let length = LENGTH_BASE[symbol] + bits.take(LENGTH_EXTRA[symbol])?;
        if length == END {
            return Ok(out);
        }
        let shift = if length == 2 {
            2
        } else {
            dictionary_bits as u32
        };
        let distance = (bits.decode(&DISTANCE)? << shift) + bits.take(shift)? + 1;
        if distance > out.len() {
            return Err(DbcError::Corrupt(bits.position));
        }
        let source = out.len() - distance;
        if distance >= length {
            out.extend_from_within(source..source + length);
        } else {
            // Overlapping copy: each new byte may be one this copy just wrote.
            for _ in 0..length {
                out.push(out[out.len() - distance]);
            }
        }
    }
}

/// DBF bytes from a DBC payload: the header is kept, the CRC32 after it is ignored.
pub fn decompress(raw: &[u8]) -> Result<Vec<u8>, DbcError> {
    let size = raw.get(8..10).ok_or(DbcError::MissingHeader)?;
    let header_size = usize::from(u16::from_le_bytes([size[0], size[1]]));
    if header_size + 4 > raw.len() {
        return Err(DbcError::HeaderSize);
    }
    let mut out = raw[..header_size].to_vec();
    out.extend(explode(raw, header_size + 4)?);
    Ok(out)
}
